### ISSUE TITLE
Immediate retry for badNonce errors

### ISSUE THREAD

**Issue by jsha** [icing/mod_md#35]
I have a mod_md that got into a bad state somehow. Right now, every time it retries, it gets a badNonce error. As far as I can tell, this triggers a retry a few minutes later, using the new nonce received with the error. Since nonces currently expire after a few minutes, this retry fails.

Bad nonce errors should be treated differently than other errors, and should be retried immediately at least once.

**Comment by icing** [comment]
I agree to the retry strategy. Could you produce a log at
```
   LogLevel md:trace2
```
of this loop behaviour?

**Comment by jsha** [comment]
```
[Tue Aug 29 21:52:39.418299 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(589): next run in  1:00:00 hours
[Tue Aug 29 22:52:47.389774 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(533): AH10055: md watchdog run, auto drive 1 mds
[Tue Aug 29 22:52:47.390141 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(479): AH10052: md(powerdns.crud.net): state=1, driving
[Tue Aug 29 22:52:47.390188 2017] [md:debug] [pid 31132:tid 140243968763648] md_reg.c(861): powerdns.crud.net: run staging
[Tue Aug 29 22:52:47.390213 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(644): powerdns.crud.net: staging started, state=1, can_http=1, can_https=1, challenges='tls-sni-01 http-01'
[Tue Aug 29 22:52:47.393829 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(680): powerdns.crud.net: checked creds
[Tue Aug 29 22:52:47.393893 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(143): get directory from https://acme-staging.api.letsencrypt.org/directory
[Tue Aug 29 22:52:47.393903 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(400): req: POST https://acme-staging.api.letsencrypt.org/directory
[Tue Aug 29 22:52:47.581377 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(418): req sent
[Tue Aug 29 22:52:47.581482 2017] [md:info] [pid 31132:tid 140243968763648] powerdns.crud.net: need certificate
[Tue Aug 29 22:52:47.585504 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_acct.c(423): acct validation
[Tue Aug 29 22:52:47.598951 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(400): req: POST https://acme-v01.api.letsencrypt.org/acme/reg/20108074
[Tue Aug 29 22:52:47.835021 2017] [md:warn] [pid 31132:tid 140243968763648] (11)Resource temporarily unavailable: acme problem urn:acme:error:badNonce: JWS has invalid anti-replay nonce N9oHDKWaXyNq6MjM_BW0KlF1jvsIqfoDAD13Ogh5J7E
[Tue Aug 29 22:52:47.835175 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(418): (11)Resource temporarily unavailable: req sent
[Tue Aug 29 22:52:47.846962 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(400): req: POST https://acme-v01.api.letsencrypt.org/acme/reg/20108074
[Tue Aug 29 22:52:48.064597 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_acct.c(402): validate acct https://acme-v01.api.letsencrypt.org/acme/reg/20108074: {\n  "id": 20108074,\n  "key": {\n    "kty": "RSA",\n    "n": "06L5iCXeRmqYzGRFOh3DW73g0vqDhXWQCeMS0OmVnULbXfgJ7d-nu6nj001lAR-GWvyXjRhBflEtagSFXq3uoOehZfvi9_oKz8U9Vq8A7b1H3FfJvCuiEy6MY7_KLqTUGXskJaDntkdtRZBeE4H7pZUFlmt6eRYH74kxHcVfldYhzVwARqO34Oj09ULYteQSJ-TzHNyTINJpmUFOd2zuFidC4vMbWqBLMN7w00WK-pR7xYrBPEDvqUrr8n7clKcQMvTEYXZD9I6dV66zGwZU_BLcwgTm0wWDfkrVFw2ZdUXeT9mZPx2_y5feoncLsP-WD7sGaE9atwj6XMXNXy-YzPPQzTcN_8sgnT6Oahlh0mrprjg5EBxD63r5kLmewrH6YmXyGvOysbM2rWbQQDRGVtBqpc4juOt_8dESm_u5_OkkAv96GybFi_U_tBIbctHI3aLZD6z6fX7MZRLFU-wozy8jioksPJQ-KlpzOw99QMragduQunCBGcZWbdBznLzcIXJVrn5oRcOfTuqUbNekXjseFbE8nRdE-zzZyOxubz_ZiTPI63HWURCrgsYSjp4Fuhax9_4_ezKvtY9PDdfHfQqz-4_lCQWZLbXXjWjMDz3TY5BjB2fis0DcB30lsv8BZTqOVlo8IKA5JYEbwoR-gdzAqkCGF21VKmRxUoyF630",\n    "e": "AQAB"\n  },\n  "contact": [\n    "mailto:jsha@newview.org"\n  ],\n  "initialIp": "2604:a880:1:20::36f9:e001",\n  "createdAt": "2017-08-18T17:23:02Z",\n  "Status": "valid"\n}
[Tue Aug 29 22:52:48.064825 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_acct.c(410): needs to agree to terms-of-service 'https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf', has already agreed to '(null)'
[Tue Aug 29 22:52:48.067058 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(418): req sent
[Tue Aug 29 22:52:48.067081 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(83): re-using staged account
[Tue Aug 29 22:52:48.067093 2017] [md:info] [pid 31132:tid 140243968763648] powerdns.crud.net: check Terms-of-Service agreement
[Tue Aug 29 22:52:48.067103 2017] [md:error] [pid 31132:tid 140243968763648] need to accept terms-of-service <https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf> for account (null)
[Tue Aug 29 22:52:48.067112 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(823): (13)Permission denied: powerdns.crud.net: ACME, check agreement
[Tue Aug 29 22:52:48.067126 2017] [md:debug] [pid 31132:tid 140243968763648] md_reg.c(868): (13)Permission denied: powerdns.crud.net: staging done
[Tue Aug 29 22:52:48.067159 2017] [md:error] [pid 31132:tid 140243968763648] (13)Permission denied: AH10056: processing powerdns.crud.net
[Tue Aug 29 22:52:48.067164 2017] [md:info] [pid 31132:tid 140243968763648] AH10057: encountered errors for the 26. time, next run in 3600 seconds
[Tue Aug 29 22:52:48.067170 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(589): next run in  1:00:00 hours
[Tue Aug 29 23:52:56.002978 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(533): AH10055: md watchdog run, auto drive 1 mds
[Tue Aug 29 23:52:56.003182 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(479): AH10052: md(powerdns.crud.net): state=1, driving
[Tue Aug 29 23:52:56.003244 2017] [md:debug] [pid 31132:tid 140243968763648] md_reg.c(861): powerdns.crud.net: run staging
[Tue Aug 29 23:52:56.003274 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(644): powerdns.crud.net: staging started, state=1, can_http=1, can_https=1, challenges='tls-sni-01 http-01'
[Tue Aug 29 23:52:56.004525 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(680): powerdns.crud.net: checked creds
[Tue Aug 29 23:52:56.004579 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(143): get directory from https://acme-staging.api.letsencrypt.org/directory
[Tue Aug 29 23:52:56.004588 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(400): req: POST https://acme-staging.api.letsencrypt.org/directory
[Tue Aug 29 23:52:56.274089 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(418): req sent
[Tue Aug 29 23:52:56.274203 2017] [md:info] [pid 31132:tid 140243968763648] powerdns.crud.net: need certificate
[Tue Aug 29 23:52:56.278582 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_acct.c(423): acct validation
[Tue Aug 29 23:52:56.292852 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(400): req: POST https://acme-v01.api.letsencrypt.org/acme/reg/20108074
[Tue Aug 29 23:52:56.572093 2017] [md:warn] [pid 31132:tid 140243968763648] (11)Resource temporarily unavailable: acme problem urn:acme:error:badNonce: JWS has invalid anti-replay nonce v8Db5KhBrhPbatV3mVeZpgw25m3RW6nObtAJYhxSqlA
[Tue Aug 29 23:52:56.572249 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(418): (11)Resource temporarily unavailable: req sent
[Tue Aug 29 23:52:56.586380 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(400): req: POST https://acme-v01.api.letsencrypt.org/acme/reg/20108074
[Tue Aug 29 23:52:56.820098 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_acct.c(402): validate acct https://acme-v01.api.letsencrypt.org/acme/reg/20108074: {\n  "id": 20108074,\n  "key": {\n    "kty": "RSA",\n    "n": "06L5iCXeRmqYzGRFOh3DW73g0vqDhXWQCeMS0OmVnULbXfgJ7d-nu6nj001lAR-GWvyXjRhBflEtagSFXq3uoOehZfvi9_oKz8U9Vq8A7b1H3FfJvCuiEy6MY7_KLqTUGXskJaDntkdtRZBeE4H7pZUFlmt6eRYH74kxHcVfldYhzVwARqO34Oj09ULYteQSJ-TzHNyTINJpmUFOd2zuFidC4vMbWqBLMN7w00WK-pR7xYrBPEDvqUrr8n7clKcQMvTEYXZD9I6dV66zGwZU_BLcwgTm0wWDfkrVFw2ZdUXeT9mZPx2_y5feoncLsP-WD7sGaE9atwj6XMXNXy-YzPPQzTcN_8sgnT6Oahlh0mrprjg5EBxD63r5kLmewrH6YmXyGvOysbM2rWbQQDRGVtBqpc4juOt_8dESm_u5_OkkAv96GybFi_U_tBIbctHI3aLZD6z6fX7MZRLFU-wozy8jioksPJQ-KlpzOw99QMragduQunCBGcZWbdBznLzcIXJVrn5oRcOfTuqUbNekXjseFbE8nRdE-zzZyOxubz_ZiTPI63HWURCrgsYSjp4Fuhax9_4_ezKvtY9PDdfHfQqz-4_lCQWZLbXXjWjMDz3TY5BjB2fis0DcB30lsv8BZTqOVlo8IKA5JYEbwoR-gdzAqkCGF21VKmRxUoyF630",\n    "e": "AQAB"\n  },\n  "contact": [\n    "mailto:jsha@newview.org"\n  ],\n  "initialIp": "2604:a880:1:20::36f9:e001",\n  "createdAt": "2017-08-18T17:23:02Z",\n  "Status": "valid"\n}
[Tue Aug 29 23:52:56.820277 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_acct.c(410): needs to agree to terms-of-service 'https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf', has already agreed to '(null)'
[Tue Aug 29 23:52:56.822334 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme.c(418): req sent
[Tue Aug 29 23:52:56.822348 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(83): re-using staged account
[Tue Aug 29 23:52:56.822354 2017] [md:info] [pid 31132:tid 140243968763648] powerdns.crud.net: check Terms-of-Service agreement
[Tue Aug 29 23:52:56.822360 2017] [md:error] [pid 31132:tid 140243968763648] need to accept terms-of-service <https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf> for account (null)
[Tue Aug 29 23:52:56.822365 2017] [md:debug] [pid 31132:tid 140243968763648] md_acme_drive.c(823): (13)Permission denied: powerdns.crud.net: ACME, check agreement
[Tue Aug 29 23:52:56.822378 2017] [md:debug] [pid 31132:tid 140243968763648] md_reg.c(868): (13)Permission denied: powerdns.crud.net: staging done
[Tue Aug 29 23:52:56.822408 2017] [md:error] [pid 31132:tid 140243968763648] (13)Permission denied: AH10056: processing powerdns.crud.net
[Tue Aug 29 23:52:56.822413 2017] [md:info] [pid 31132:tid 140243968763648] AH10057: encountered errors for the 27. time, next run in 3600 seconds
[Tue Aug 29 23:52:56.822419 2017] [md:debug] [pid 31132:tid 140243968763648] mod_md.c(589): next run in  1:00:00 hours
```

**Comment by jsha** [comment]
Whoops, that was only debug. I set it to trace2 and restarted, and collected the following log:

[trace.txt](https://github.com/icing/mod_md/files/1262170/trace.txt)

BTW, does mod_md always reset the retry time when Apache restarts?

**Comment by icing** [comment]
It does. Restarts in production should be relatively rare, I assumed.

**Comment by icing** [comment]
The badNonce already led to a retry but was too verbose and confused reading the log. Especially since badNonce is happening every time on the first request.

The real issue why the log above shows a stuck MD is that the Agreement did not match/was unavailable. I updated the message in the log to better reflect what the user is expected to do.

Also, missing/mismatched agreements no longer trigger error retries, since retries do not change anything in this case. 

The fault were a changed agreement does not have any effect is reported in issue #31 so I intend to close this issue, if you have no objections.

**Comment by icing** [comment]
Could you check your case if v0.8.0 improves the reporting? Thanks!

**Comment by jsha** [comment]
This looks much better:

```
[Thu Aug 31 18:51:34.302307 2017] [md:error] [pid 14901:tid 140187961648896] (70008)Partial results are valid but processing is incomplete: powerdns.crud.net: the CA requires you to accept the terms-of-service as specified in <https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf>.\nPlease read the document that you find at that URL and, if you agree to the condictions, configure \n  MDCertificateAgreement url\nwith exactly that URL in your Apache. Then restart the server to activate.
```

Note that the `\n` shows up literally in the log output, which I'm guessing is a mistake. Also, is a reload sufficient, or is a full restart required? If a reload is enough, it would be better to say that here.

**Comment by icing** [comment]
Thanks. Will change the text slightly.

**Comment by icing** [comment]
Available in v0.8.2.
